cpp/src/arrow/compute/exec/union_node.cc

```diff
@@ -48,31 +48,29 @@ struct UnionNode : ExecNode {
       : ExecNode(plan, inputs, GetInputLabels(inputs),
                  /*output_schema=*/inputs[0]->output_schema(),
                  /*num_outputs=*/1) {
-    if (this->input_count_.SetTotal(static_cast<int>(inputs.size()))) {
-      finished_.MarkFinished();
-    }
+    ARROW_DCHECK(this->input_count_.SetTotal(static_cast<int>(inputs.size())) == false);
```

Comment:
This won't work - this will get compiled out in release mode.

Reply:
In particular I think this is responsible for the MinGW CI failures (which compile in release mode vs the others which compile in debug mode)